common/archiver/gcloud/historyArchiver.go

```diff
@@ -0,0 +1,387 @@
+
+		historyBatches = historyBatches[token.NextBatchIdx:]
+
+		for _, batch := range historyBatches {
+			response.HistoryBatches = append(response.HistoryBatches, batch)
+			numOfBatches++
+			numOfEvents += len(batch.Events)
+			if numOfEvents >= request.PageSize {
+				pageSizeCompleted = true
+				break
+			}
+		}
+
+		if numOfBatches < len(historyBatches) {
+			token.NextBatchIdx += numOfBatches
+			nextToken, err := serializeToken(token)
+			if err != nil {
+				return nil, &shared.InternalServiceError{Message: err.Error()}
+			}
+			response.NextPageToken = nextToken
+		}
+	}
+
+	return response, nil
+
+}
+
+// ValidateURI is used to define what a valid URI for an implementation is.
+func (h *historyArchiverData) ValidateURI(URI archiver.URI) (err error) {
+	if URI.Scheme() != URIScheme {
+		return archiver.ErrURISchemeMismatch
+	}
+
+	if URI.Path() == "" || URI.Hostname() == "" {
+		return archiver.ErrInvalidURI
+	}
+
+	return
+}
+
+func (h *historyArchiverData) validateURI(ctx context.Context, URI archiver.URI) (err error) {
+	if err = h.ValidateURI(URI); err == nil {
+		_, err = h.gcloudStorage.Exist(ctx, URI, "")
+	}
+
+	return
+}
+
+func getNextHistoryBlob(ctx context.Context, historyIterator archiver.HistoryIterator) (*archiver.HistoryBlob, error) {
+	historyBlob, err := historyIterator.Next()
+	op := func() error {
+		historyBlob, err = historyIterator.Next()
+		return err
+	}
+	for err != nil {
+		if !common.IsPersistenceTransientError(err) {
+			return nil, err
+		}
+		if contextExpired(ctx) {
+			return nil, archiver.ErrContextTimeout
+		}
+		err = backoff.Retry(op, common.CreatePersistanceRetryPolicy(), common.IsPersistenceTransientError)
+	}
+	return historyBlob, nil
+}
+
+func historyMutated(request *archiver.ArchiveHistoryRequest, historyBatches []*shared.History, isLast bool) bool {
+	lastBatch := historyBatches[len(historyBatches)-1].Events
+	lastEvent := lastBatch[len(lastBatch)-1]
+	lastFailoverVersion := lastEvent.GetVersion()
+	if lastFailoverVersion > request.CloseFailoverVersion {
+		return true
+	}
+
+	if !isLast {
+		return false
+	}
+	lastEventID := lastEvent.GetEventId()
+	return lastFailoverVersion != request.CloseFailoverVersion || lastEventID+1 != request.NextEventID
+}
+
+func (h *historyArchiverData) getHighestVersion(ctx context.Context, URI archiver.URI, request *archiver.GetHistoryRequest) (*int64, []int64, error) {
+
+	filenames, err := h.gcloudStorage.Query(ctx, URI, constructHistoryFilenamePrefix(request.DomainID, request.WorkflowID, request.RunID))
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var highestVersion *int64
+	batchesPartIds := make([]int64, 0, 0)
+	for _, filename := range filenames {
+		version, partVersionId, err := extractCloseFailoverVersion(filepath.Base(filename))
+		if err != nil {
+			continue
+		}
+		if highestVersion == nil || version > *highestVersion {
+			highestVersion = &version
+		}
+
+		if highestVersion == nil || *highestVersion == version {
```

Comment:
I don't think highestVersion will be `nil` here.

Reply:
agree!